ledger-core/virtual/integration/vstatereport_test.go

```diff
@@ -92,3 +100,502 @@ func TestVirtual_SendVStateReport_IfPulseChanged(t *testing.T) {
+		require.False(t, outgoingRef.IsZero())
+		require.False(t, outgoingRef.IsEmpty())
+
+		found := false
+		for _, reqInfo := range s.requests {
+			if outgoingRef.Equal(reqInfo.ref) {
+				found = true
+				reqInfo.token = del.DelegationSpec
+				break
+			}
+		}
+		require.True(t, found)
+		return false
+	})
+	typedChecker.VDelegatedCallRequest.Set(func(req *payload.VDelegatedCallRequest) bool {
+		outgoingRef := req.CallOutgoing
+		require.False(t, outgoingRef.IsZero())
+		require.False(t, outgoingRef.IsEmpty())
+
+		require.Equal(t, s.getObject(), req.Callee)
+
+		pl := payload.VDelegatedCallResponse{
+			DelegationSpec: payload.CallDelegationToken{
+				TokenTypeAndFlags: payload.DelegationTokenTypeCall,
+				PulseNumber:       s.getPulse(5),
+				Callee:            s.getObject(),
+				Outgoing:          outgoingRef,
+				ApproverSignature: []byte("deadbeef"),
+			},
+		}
+		s.addPayload(ctx, &pl)
+		return false
+	})
+	typedChecker.VDelegatedRequestFinished.Set(func(res *payload.VDelegatedRequestFinished) bool {
+		outgoingRef := res.CallOutgoing
+		require.False(t, outgoingRef.IsZero())
+		require.False(t, outgoingRef.IsEmpty())
+
+		require.Equal(t, s.getObject(), res.Callee)
+
+		return false
+	})
+	typedChecker.VCallResult.Set(func(res *payload.VCallResult) bool {
+		outgoingRef := res.CallOutgoing
+		require.False(t, outgoingRef.IsZero())
+		require.False(t, outgoingRef.IsEmpty())
+
+		require.Equal(t, s.getObject(), res.Callee)
+
+		return false
+	})
+}
+
+func (s *stateReportCheckPendingCountersAndPulsesTest) getPulse(
+	pulseIndex int,
+) pulse.Number {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	return s.pulses[pulseIndex]
+}
+
+func (s *stateReportCheckPendingCountersAndPulsesTest) getCurrentPulse() pulse.Number {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	return s.pulses[s.currentPulseIndex]
+}
+
+func (s *stateReportCheckPendingCountersAndPulsesTest) getObject() reference.Global {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	return s.object
+}
+
+func (s *stateReportCheckPendingCountersAndPulsesTest) getCaller() reference.Global {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	return s.caller
+}
+
+func (s *stateReportCheckPendingCountersAndPulsesTest) waitMessagePublications(
+	ctx context.Context,
+	t *testing.T,
+	expected int,
+) {
+	if !s.server.PublisherMock.WaitCount(expected, 10*time.Second) {
+		panic("timeout waiting for messages on publisher")
+	}
+	require.Equal(t, expected, s.server.PublisherMock.GetCount())
+}
+
+func (s *stateReportCheckPendingCountersAndPulsesTest) addPayload(
+	ctx context.Context, pl payload.Marshaler,
+) {
+	me := s.server.JetCoordinatorMock.Me()
+	msg := utils.NewRequestWrapper(s.getCurrentPulse(), pl).SetSender(me).Finalize()
+	s.server.SendMessage(ctx, msg)
+}
+
+func (s *stateReportCheckPendingCountersAndPulsesTest) addPayloadAndWaitIdle(
+	ctx context.Context, pl payload.Marshaler,
+) {
+	s.server.SuspendConveyorAndWaitThenResetActive()
+	s.addPayload(ctx, pl)
+	s.server.WaitActiveThenIdleConveyor()
```

Comment:
why?

Reply:
to get deterministic order of operations as much as I can

Reply:
discussed this